Homework 6 7

EremitePK opened the pull request:
Исправил все по комментариям. 
По комментариям:
1. Удобно будет сделать requireNotNull. Сделал.
2. Можно перенести чутка кода в Utils. СДЕЛАЛ
3. ? - тут не нужен
так как actorsAdaptor не nullable
студия тебе подсказывает это подсвечивая код. СДЕЛАЛ
4. SpanCount - SPAN_COUNT константы пишем большими буквами. УДАЛЕН
5. мы далаем копию ложим ее в currentMovie
и дальше ничего с этой переменной не делаем
зачем так?) ИСПРАВИЛ
6. формат строки на число исправил
7. удалить - ? RecyclerViewAdapter ИСПРАВИЛ
8. нужно проверять по id DiffUtilCallbackSolution.kt ИСПРАВИЛ
9. 

Попытался реализовать MVVM. Помоему не получилось. Я скорей всего чего то не понимаю. 
Посмотри пожалуйста направь где смотреть. А то я уже просто не знаю как и где неправильно делаю. Я почему то не могу отследить разделение на части. Заранее извеняюсь за код. Там просто жесть. Но сам я уже придумать ни чего не могу. Запутался. 
Загрузка с сети популярных фильмов работает. Также актеры подгружаются и жанры.

EremitePK committed:
Исправил все по комментариям. Попытался реализовать MVVM. Помоему не получилось. Я скорей всего чегото не понимаю. Посмотри пожалуйста направь где смотреть. А то я уже просто не знаю как и где неправильно делаю. Я почемуто не могу отследить разделение на части. Заранее извеняюсь за код. Там просто жесть. Но сам я уже придумать ни чего не могу. Запутался. Загрузка с сети популярных фильмов работает. Также актеры подгружаются и жанры.
Создал отдельные классы для отделения модели от вью. (PresentetionModelData).
Иправил по коментариям:
1. FragmentMovieDetails.kt: не нужно делать movie переменной класса ее просто можно сразу же показать ИСПРАВИЛ
2. poster -> posterImageView исправлено
3. уже вроде как сontext и не нужен в фабрики ИСПРАВИЛ
4. RecyclerViewAdapter наверное стоит переименовать. Это кусок от попытки написания универсального адаптера реализованый полностью остался в Homework 5_1
5. ClassUtils.kt: Удален
6. ViewModel и Interactor разделены
7. Persistent - это про хранения ПЕРЕИМЕНОВАН в Repository
8. API Network перенесены в отдельный класс
9. IDMOVIE переменные так не называем ИСПРАВИЛ
10. Было бы круто если RetrofitModule.genreApi приходило извне ЭТОГО НЕ ПОНЯЛ
11. CoroutineScope(Dispatchers.Main) -> viewModelScope ИСПРАВЛЕНО
Исправил получение информации по возрастному ограничению и длительности фильма (из за того что приходится дополнительно делать +2 запроса для получения этой информации скорость загрузки стала гораздо хуже). И немного поправил интерфейс и удалил лишнее.
Пришлось сделать подгрузку возростного ограничения и продолжительности после загрузки основных данных и передачи их IO чтоб пользователь долго не ждал загрузки всех данных.
Равлизовал функцию открытия окна со списком фильмов при нажатии кнопки андроид back. Также сделал чтоб при повороте экрана сохранялась позиция фильма в списке.
Реализовал работу кнопки Like (помоему не правильно но работает). Также все переменные переживают перевороты  и смену окна

EremitePK committed:
Исправил все по комментариям. Попытался реализовать MVVM. Помоему не получилось. Я скорей всего чегото не понимаю. Посмотри пожалуйста направь где смотреть. А то я уже просто не знаю как и где неправильно делаю. Я почемуто не могу отследить разделение на части. Заранее извеняюсь за код. Там просто жесть. Но сам я уже придумать ни чего не могу. Запутался. Загрузка с сети популярных фильмов работает. Также актеры подгружаются и жанры.

EremitePK committed:
Исправил все по комментариям. Попытался реализовать MVVM. Помоему не получилось. Я скорей всего чегото не понимаю. Посмотри пожалуйста направь где смотреть. А то я уже просто не знаю как и где неправильно делаю. Я почемуто не могу отследить разделение на части. Заранее извеняюсь за код. Там просто жесть. Но сам я уже придумать ни чего не могу. Запутался. Загрузка с сети популярных фильмов работает. Также актеры подгружаются и жанры.
Создал отдельные классы для отделения модели от вью. (PresentetionModelData).
Иправил по коментариям:
1. FragmentMovieDetails.kt: не нужно делать movie переменной класса ее просто можно сразу же показать ИСПРАВИЛ
2. poster -> posterImageView исправлено
3. уже вроде как сontext и не нужен в фабрики ИСПРАВИЛ
4. RecyclerViewAdapter наверное стоит переименовать. Это кусок от попытки написания универсального адаптера реализованый полностью остался в Homework 5_1
5. ClassUtils.kt: Удален
6. ViewModel и Interactor разделены
7. Persistent - это про хранения ПЕРЕИМЕНОВАН в Repository
8. API Network перенесены в отдельный класс
9. IDMOVIE переменные так не называем ИСПРАВИЛ
10. Было бы круто если RetrofitModule.genreApi приходило извне ЭТОГО НЕ ПОНЯЛ
11. CoroutineScope(Dispatchers.Main) -> viewModelScope ИСПРАВЛЕНО
Исправил получение информации по возрастному ограничению и длительности фильма (из за того что приходится дополнительно делать +2 запроса для получения этой информации скорость загрузки стала гораздо хуже). И немного поправил интерфейс и удалил лишнее.
Пришлось сделать подгрузку возростного ограничения и продолжительности после загрузки основных данных и передачи их IO чтоб пользователь долго не ждал загрузки всех данных.
Равлизовал функцию открытия окна со списком фильмов при нажатии кнопки андроид back. Также сделал чтоб при повороте экрана сохранялась позиция фильма в списке.
Реализовал работу кнопки Like (помоему не правильно но работает). Также все переменные переживают перевороты  и смену окна.Сделал кэш

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/FragmentMovieDetails.kt:
```diff
@@ -8,96 +8,138 @@ import android.view.ViewGroup
 class FragmentMovieDetails : Fragment() {
+    private val viewModel: MoviesViewModel by viewModels { MoviesViewModelFactory(requireContext()) }
     private var listener: TopMainMenuClickListener? = null
+    private var movie: PresentationModelData.MoviePresentation? = null
```
не нужно делать  movie переменной класса
ее просто можно сразу же показать

ivanlardis replied:
![image](https://user-images.githubusercontent.com/11761610/104195225-63d68300-5433-11eb-935d-5100f4844ff9.png)
Плюс минус так было бы лучше
но пугает количество подписок
Тут для деталки один запрос
Для списка другой так что нужн было две VM
на каждый фрагмент свой!

EremitePK replied:
Разделил вьюмодели по фрагментам

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/FragmentMovieDetails.kt:
```diff
@@ -8,96 +8,138 @@ import android.view.ViewGroup
         super.onDetach()
     }
+
     override fun onStart() {
```
удаляем неиспользуемое)
такое грязнит код)

ivanlardis replied:
done

EremitePK replied:
Исправил

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/FragmentMovieDetails.kt:
```diff
@@ -8,96 +8,138 @@ import android.view.ViewGroup
+        (actorsAdaptor).bindActors(listActors)
+    }
+
+    private fun updateData() {
```
наверное стоит удалить этот метод

EremitePK replied:
Удалил

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/FragmentMovieDetails.kt:
```diff
@@ -8,96 +8,138 @@ import android.view.ViewGroup
+    private var idMovie: Int? = null
+    private lateinit var actorsAdaptor: ActorsAdapter
+    private var poster: ImageView? = null
+    private var age: TextView? = null
```
poster -> posterImageView
чтоб из кода было понятно с чем работаем)

EremitePK replied:
Сделал

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/FragmentMoviesList.kt:
```diff
@@ -2,83 +2,100 @@ package ru.eremite.myapplication
 
 class FragmentMoviesList : Fragment() {
 
+    private val viewModel: MoviesViewModel by viewModels { MoviesViewModelFactory(requireContext()) }
```
уже вроде как сontext и не нужен в фабрики

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/RecyclerViewAdapter.kt:
```diff
@@ -9,18 +9,17 @@ import android.widget.TextView
+import ru.eremite.myapplication.data.Header
+import ru.eremite.myapplication.data.PresentationModelData
 
 class RecyclerViewAdapter(
```
RecyclerViewAdapter наверное стоит переименовать)

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/ClassUtils.kt:
```diff
@@ -1,11 +1,19 @@
         return "http://lardis.ru/academ/webp/$failName.webp"
     }
+
+    suspend fun loadData(context: Context): List<ModelData.Movie> {
```
этот класс можно удалить)

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/MovieInteractor.kt:
```diff
@@ -0,0 +1,16 @@
+    private val persistent: MoviePersistentNetwork
+) {
+    suspend fun loadMovies(): List<ModelData.Movie> = persistent.loadMovies()
+    fun findMovies(idMovie: Int, listMovies: List<ModelData.Movie>): ModelData.Movie? =
```
findMovies не используется
раз будут две VM значит будет
и два   интерактора

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/MoviePersistentNetwork.kt:
```diff
@@ -0,0 +1,77 @@
+import ru.eremite.myapplication.data.ModelData
+import ru.eremite.myapplication.data.MoviesResponse
+
+class MoviePersistentNetwork : MoviePersistent() {
```
Persistent - это про хранения
Network - про подход в сеть

ivanlardis replied:
Тоесть потом будет два класса 
MovieNetwork
MoviePersistent или MovieStorage

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/MoviePersistentNetwork.kt:
```diff
@@ -0,0 +1,77 @@
+    val actorsApi: ActorsAPI = retrofit.create()
+}
+
+private interface MovieApi {
```
![image](https://user-images.githubusercontent.com/11761610/104196697-2541c800-5435-11eb-8b2d-c7a514993fa1.png)
Лучше все это перенести в отдельный один класс

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/MoviePersistentNetwork.kt:
```diff
@@ -0,0 +1,77 @@
+
+private interface ActorsAPI {
+    @GET(BaseURL.BASE_ACTORS)
+    suspend fun getActors(@Path("IDMOVIE") IDMOVIE: Int): ActorsResponse
```
IDMOVIE переменные так не называем

ivanlardis replied:
![image](https://user-images.githubusercontent.com/11761610/104196905-689c3680-5435-11eb-8dff-86e40127c55a.png)

EremitePK replied:
Исправил

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/MovieViewModelFactory.kt:
```diff
@@ -0,0 +1,19 @@
+import androidx.lifecycle.ViewModelProvider
+import kotlinx.coroutines.Dispatchers
+
+class MoviesViewModelFactory(private val context: Context) : ViewModelProvider.Factory {
```
context неиспользуется

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/MoviePersistentNetwork.kt:
```diff
@@ -0,0 +1,77 @@
+import ru.eremite.myapplication.data.ModelData
+import ru.eremite.myapplication.data.MoviesResponse
+
+class MoviePersistentNetwork : MoviePersistent() {
```
Было бы круто если 
RetrofitModule.genreApi
приходило извне

EremitePK replied:
Это если честно не могу понять.

EremitePK replied:
Исправил

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/MoviesViewModel.kt:
```diff
@@ -0,0 +1,90 @@
+import kotlinx.coroutines.launch
+import ru.eremite.myapplication.data.PresentationModelData
+
+class MoviesViewModel(private val interactor: MovieInteractor) : ViewModel() {
```
разделить на две VM

EremitePK replied:
СДЕЛАЛ

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/MoviesViewModel.kt:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    private fun loadMovies() {
+        CoroutineScope(Dispatchers.Main).launch {
```
CoroutineScope(Dispatchers.Main) -> viewModelScope

EremitePK replied:
Сделал

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/FragmentMovieDetails.kt:
```diff
@@ -8,72 +8,71 @@ import android.view.ViewGroup
         savedInstanceState: Bundle?
     ): View? {
         // Inflate the layout for this fragment
         var view = inflater.inflate(R.layout.fragment_movie_details, container, false)
```
val view
студия подчеркивает даже
![image](https://user-images.githubusercontent.com/11761610/105579276-a171cf00-5d96-11eb-82c1-f950acec81ce.png)

EremitePK replied:
Исправил

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/FragmentMovieDetails.kt:
```diff
@@ -8,72 +8,71 @@ import android.view.ViewGroup
+            ViewModelProvider(requireActivity(), MovieDetailsViewModelFactory()).get(
+                MovieDetailsViewModel::class.java
+            )
+        movieViewModel.loadingCurrentMovie.observe(
```
![image](https://user-images.githubusercontent.com/11761610/105579339-fad9fe00-5d96-11eb-8994-44c51fd0e1fe.png)
Так бы получше бы выглядело

EremitePK replied:
Исправил

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/ViewModel/MovieDetailsViewModel.kt:
```diff
@@ -0,0 +1,69 @@
+import kotlinx.coroutines.launch
+
+class MovieDetailsViewModel(private val interactor: MovieDetailsInteractor) : ViewModel() {
+    private val viewModelScope = CoroutineScope(Dispatchers.Main)
```
viewModelScope нужно удалить


так как viewModelScope есть у вьюмодели
нужно использовать его

EremitePK replied:
Не понял

EremitePK replied:
Исправил

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/ViewModel/MovieDetailsViewModel.kt:
```diff
@@ -0,0 +1,69 @@
+            mutableLoadingCurrentMovie.value = true
+            val findMovie = interactor.findMovies(idMovie)
+            findMovie?.let {
+                mutableCurrentMovie.value = PresentationModelData.MoviePresentation(
```
![image](https://user-images.githubusercontent.com/11761610/105583581-3da0d380-5da1-11eb-9ebb-89da41127537.png)
![image](https://user-images.githubusercontent.com/11761610/105583633-41345a80-5da1-11eb-902f-e6b4c95295e2.png)
Конвертацию лучше вынести в отдельный метод

EremitePK replied:
Выделил в презинтейшнмодель метод

EremitePK replied:
Создал отдельный класс маппер

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/ViewModel/MovieDetailsViewModel.kt:
```diff
@@ -0,0 +1,69 @@
+                    it.getGeners()
+                )
+            }
+            mutableCurrentMovie.value?.let {
```
тут не очень хорошо что
мы дергаем второй запрос после выполнения первого
Лучше бы из паралельно выполняли в  нетворке
А тут уже получали склеиный ответ актеров и видео

EremitePK replied:
Не сделал. Моя ошибка исправлюсь.

EremitePK replied:
Если честно не могу сообразить как сделать.

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/ViewModel/MovieDetailsViewModel.kt:
```diff
@@ -0,0 +1,69 @@
+            var listActorsPresint: MutableList<PresentationModelData.ActorPresentation> =
+                mutableListOf()
+            listActors.forEach {
+                listActorsPresint.add(
```
![image](https://user-images.githubusercontent.com/11761610/105585289-b6a02b00-5da1-11eb-9a1b-a971b7ddebe2.png)
forEach add
Можно и нужно поменять на map
и 
  PresentationModelData.ActorPresentation(
                        it.id,
                        it.name,
                        it.picture
                    )
перенсти в метод по анологии выше

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/ViewModel/MoviesListViewModel.kt:
```diff
@@ -0,0 +1,140 @@
+        }
+    }
+
+    fun findMoviesByName(name: String, notViewLoader: Boolean = false) {
```
очень большой метод нужно его разделить на несколько методов

EremitePK replied:
Не разделил.

EremitePK replied:
Не вышло пока разделить но додумаю позже

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/data/PopularMoviesResponse.kt:
```diff
@@ -0,0 +1,208 @@
+            this.results = results
+        }
+
+    fun setResults(results: List<MovieNetwork>?) {
```
ненужный метод
в колтине нет сетеров

EremitePK replied:
Исправил

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/MovieRepositoryNetwork.kt:
```diff
@@ -0,0 +1,107 @@
+import retrofit2.http.Query
+import ru.eremite.myapplication.data.*
+
+class MovieRepositoryNetwork : MovieRepository() {
```
Должны быть классы 
MovieRepository
MovieNetwork

MovieRepositoryNetwork такого не должно быть
https://github.com/AndroidArchitecture/AndroidArchitectureBook
Тут можно почитать об этом

EremitePK replied:
Почитал. И если честно что то я наверно совсем

EremitePK replied:
Исправил. Выделил классы как положено

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/MovieRepositoryNetwork.kt:
```diff
@@ -0,0 +1,107 @@
+
+class MovieRepositoryNetwork : MovieRepository() {
+    override suspend fun loadMovies(currentPage: Int): List<ModelData.Movie> {
+        val genres = BaseDataTMDb.networkApi.getGenre().genresList
```
networkApi
тоже должно приходить через конструктор

EremitePK replied:
Тоже не могу понять.

EremitePK replied:
иСПРАВИЛ

ivanlardis commented on app/src/main/java/ru/eremite/myapplication/utils/MovieRepositoryNetwork.kt:
```diff
@@ -0,0 +1,107 @@
+    }
+}
+
+object BaseDataTMDb {
```
MovieRepositoryNetwork
BaseDataTMDb
Api
Принято разносить по разным классам

EremitePK replied:
И это не могу понять

EremitePK replied:
Сделал

EremitePK commented:
Реализовал работу с кэшем роом. Через flow/

EremitePK committed:
Исправил все по комментариям. Попытался реализовать MVVM. Помоему не получилось. Я скорей всего чегото не понимаю. Посмотри пожалуйста направь где смотреть. А то я уже просто не знаю как и где неправильно делаю. Я почемуто не могу отследить разделение на части. Заранее извеняюсь за код. Там просто жесть. Но сам я уже придумать ни чего не могу. Запутался. Загрузка с сети популярных фильмов работает. Также актеры подгружаются и жанры.
Создал отдельные классы для отделения модели от вью. (PresentetionModelData).
Иправил по коментариям:
1. FragmentMovieDetails.kt: не нужно делать movie переменной класса ее просто можно сразу же показать ИСПРАВИЛ
2. poster -> posterImageView исправлено
3. уже вроде как сontext и не нужен в фабрики ИСПРАВИЛ
4. RecyclerViewAdapter наверное стоит переименовать. Это кусок от попытки написания универсального адаптера реализованый полностью остался в Homework 5_1
5. ClassUtils.kt: Удален
6. ViewModel и Interactor разделены
7. Persistent - это про хранения ПЕРЕИМЕНОВАН в Repository
8. API Network перенесены в отдельный класс
9. IDMOVIE переменные так не называем ИСПРАВИЛ
10. Было бы круто если RetrofitModule.genreApi приходило извне ЭТОГО НЕ ПОНЯЛ
11. CoroutineScope(Dispatchers.Main) -> viewModelScope ИСПРАВЛЕНО
Исправил получение информации по возрастному ограничению и длительности фильма (из за того что приходится дополнительно делать +2 запроса для получения этой информации скорость загрузки стала гораздо хуже). И немного поправил интерфейс и удалил лишнее.
Пришлось сделать подгрузку возростного ограничения и продолжительности после загрузки основных данных и передачи их IO чтоб пользователь долго не ждал загрузки всех данных.
Равлизовал функцию открытия окна со списком фильмов при нажатии кнопки андроид back. Также сделал чтоб при повороте экрана сохранялась позиция фильма в списке.
Реализовал работу кнопки Like (помоему не правильно но работает). Также все переменные переживают перевороты  и смену окна
